test/e2e/btc_rewards_distribution_ibc_callbacks_e2e_test.go

```diff
@@ -0,0 +1,456 @@
+
+	babylonFps := bbnNode.QueryFinalityProviders(bbnNode.ChainID())
+	cons0Fps := bbnNode.QueryFinalityProviders(bsn0.ConsumerId)
+
+	require.Len(s.T(), append(babylonFps, cons0Fps...), 3, "should have created all the FPs to start the test")
+	s.T().Log("All Fps created")
+
+	bbnNode.WaitForNextBlock()
+}
+
+// Test2CreateBtcDelegations creates 3 btc delegations
+func (s *IbcCallbackBsnAddRewards) Test2CreateBtcDelegations() {
+	bbnNode := s.BbnNode()
+
+	s.del1Addr = bbnNode.KeysAdd(wDel1)
+	s.del2Addr = bbnNode.KeysAdd(wDel2)
+
+	bbnNode.BankMultiSendFromNode([]string{s.del1Addr, s.del2Addr}, "1000000ubbn")
+
+	bbnNode.WaitForNextBlock()
+
+	s.CreateBTCDelegationMultipleFPsAndCheck(bbnNode, wDel1, s.del1BTCSK, s.del1Addr, s.fp2Del1StkAmt, s.fp1bbn, s.fp2cons0)
+	s.CreateBTCDelegationMultipleFPsAndCheck(bbnNode, wDel1, s.del1BTCSK, s.del1Addr, s.fp3Del1StkAmt, s.fp1bbn, s.fp3cons0)
+	s.CreateBTCDelegationMultipleFPsAndCheck(bbnNode, wDel2, s.del2BTCSK, s.del2Addr, s.fp2Del2StkAmt, s.fp1bbn, s.fp2cons0)
+
+	resp := bbnNode.QueryBtcDelegations(bstypes.BTCDelegationStatus_ANY)
+	require.Len(s.T(), resp.BtcDelegations, 3, "should have all 3 delegations")
+
+	s.CreateCovenantsAndSubmitSignaturesToPendDels(bbnNode, s.fp1bbn)
+
+	s.bbnIbcCallbackReceiverAddr = bbnNode.KeysAdd("bsn-receiver")
+	s.T().Log("All BTC delegations created")
+}
+
+func (s *IbcCallbackBsnAddRewards) Test3CreateFactoryToken() {
+	bsnNode := s.BsnNode()
+
+	// Create and fund sender account
+	s.bsnSenderAddr = bsnNode.KeysAdd("bsn-sender")
+	bsnNode.BankSendFromNode(s.bsnSenderAddr, "15000000ubbn") // Give enough ubbn for tokenfactory creation fee (10M) + tx fees
+	bsnNode.WaitForNextBlock()
+
+	// Create custom denom using tokenfactory
+	customDenomName := datagen.GenRandomHexStr(s.r, 10)
+	s.bsnCustomTokenDenom = fmt.Sprintf("factory/%s/%s", s.bsnSenderAddr, customDenomName)
+	s.T().Logf("Creating custom denom: %s", s.bsnCustomTokenDenom)
+
+	bsnNode.CreateDenom(s.bsnSenderAddr, customDenomName)
+	bsnNode.WaitForNextBlock()
+
+	mintAmt := s.r.Int63n(10_000000) + 10_000000
+	mintInt := math.NewInt(mintAmt)
+
+	bsnNode.MintDenom(s.bsnSenderAddr, mintInt.String(), s.bsnCustomTokenDenom)
+	bsnNode.WaitForNextBlock()
+
+	bsnSenderBalances, err := bsnNode.QueryBalances(s.bsnSenderAddr)
+	s.Require().NoError(err)
+
+	// Check custom denom balance specifically
+	customBalance := bsnSenderBalances.AmountOf(s.bsnCustomTokenDenom)
+	require.Equal(s.T(), customBalance.String(), mintInt.String(), "Should have custom tokens after minting")
+}
+
+func (s *IbcCallbackBsnAddRewards) Test4SendBsnRewardsCallback() {
+	bbnNode := s.BbnNode()
+	bsnNode := s.BsnNode()
+
+	transferAmt := s.r.Int63n(2_000000) + 1_000000
+	tranferInt := math.NewInt(transferAmt)
+	rewardCoin := sdk.NewCoin(s.bsnCustomTokenDenom, tranferInt)
+
+	fp2Ratio, fp3Ratio := math.LegacyMustNewDecFromStr("0.7"), math.LegacyMustNewDecFromStr("0.3")
+
+	// Create JSON callback memo for IBC callback middleware with BSN action
```

Comment:
Important part where it sends the ibc transfer with this memo and checks the rewards reached the other side accordingly